File: 21557 - Mukesh J/Week-5/785-is-graph-bipartite/785-is-graph-bipartite.java
class Solution {
    public boolean bfs(int index, int[][] graph, int[] state)
    {
        Queue<Integer> adjacent = new LinkedList<>();
        adjacent.offer(index);
        state[index]=0;
        int node;
        while(!adjacent.isEmpty())
        {
            node = adjacent.poll();
            for(int j=0;j<graph[node].length;j++)
            {
                if(state[graph[node][j]]==-1)
                {
                    adjacent.offer(graph[node][j]);
                    state[graph[node][j]] = 1- state[node];
                }
                else if(state[node]==state[graph[node][j]])
                    return false;
            }
        }
        return true;
    }
    
    public boolean isBipartite(int[][] graph) {
        int[] state = new int[graph.length];
        Arrays.fill(state,-1);
        for(int i=0;i<graph.length;i++)
        {
            if(state[i]==-1)
                if(bfs(i,graph,state)==false)
                    return false;
        }
        return true;
    }
}

File: 21553 - Sai Nandan D/Week 5/785_Is Graph Bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        colored = {i : 0 for i in range(n)}
        
        for node in range(n):
            if not colored[node]:
                colored[node] = 1
                q = deque([node])
                while q:
                    cur_node = q.popleft()
                    for nbr in graph[cur_node]:
                        if not colored[nbr]:
                            colored[nbr] = -colored[cur_node]
                            q.append(nbr)
                        elif colored[nbr] == colored[cur_node]:
                            return False                                    
        
        return True
